pkgs/tools/admin/google-cloud-sdk/default.nix

```diff
@@ -1,23 +1,33 @@
-{ stdenv, lib, fetchurl, python, cffi, cryptography, pyopenssl, crcmod, google-compute-engine, makeWrapper }:
+# The package google-compute-engine is used by this package only on GCE
+# instances (see https://github.com/NixOS/nixpkgs/issues/31679). The flag
+# `forGceInstance` controls whether google-compute-engine should be made
+# a dependency and is disabled by default. The default NixOS configuration
+# for GCE located in nixos/modules/virtualisation/google-compute-image.nix
+# overrides this package making `forGceInstance` enabled by default.
+
+{ stdenv, lib, fetchurl, makeWrapper,
+  python, cffi, cryptography, pyopenssl, crcmod, google-compute-engine,
+  forGceInstance ? false }:
 
-# other systems not supported yet
 let
-  pythonInputs = [ cffi cryptography pyopenssl crcmod google-compute-engine ];
+  gcePackages = if forGceInstance then [ google-compute-engine ] else [];
+  pythonInputs = [ cffi cryptography pyopenssl crcmod ] ++ gcePackages;
```

Comment:
I'd rather see `[ cffi cryptography pyopenssl crcmod ] ++ lib.optional (!stdenv.isDarwin) google-compute-engine` than introduce overrides and overlays. I don't think that the closure size matters that much in this context.

Reply:
Well, in this case installing `google-cloud-sdk` from `nixpkgs` would install `systemd` as it is dependency to `google-compute-engine`. IMO, it is hardly a sane default, especially for non-NixOS users.

Reply:
I reiterate that the package `google-compute-engine` (the original name is [compute-image-packages](https://github.com/GoogleCloudPlatform/compute-image-packages)) is only meant for GCE guests. Installing it everywhere just in case seems to much to me.

Reply:
In that case let's make two top-level packages, `google-cloud-sdk` and `google-cloud-sdk-gce`. Changing the overlay of the GCE image profile seems a bit too much.

Reply:
Can't we just add `(google-cloud-sdk.override { gceClient = true; }` into `environment.systemPackages`? I think that covers most of SDK usage without much complexity, and we can skip adding an additional package too (don't have a strong opinion here).

There are also potential users of SDK as a library like TensorFlow -- I think it supports transparently using Google Cloud Storage when it can find both `google-cloud-sdk` and `google-cloud-engine` but I didn't try that; let's leave this usecase be for a time.

Reply:
@abbradar If one puts overriden `google-cloud-sdk` in `systemPackages` it won't change the fact that all packages using `google-cloud-sdk` (e.g. TensorFlow) will use the default version (without gce).

An overlay for GCE would override the default version of `google-cloud-sdk` and this is exactly why I added it. The packages built against `pkgs` with this overlay (e.g. on a default NixOS GCE image) would then use gce-enabled `google-cloud-sdk`. The problem is that all packages that use `google-cloud-sdk` need to be built with two different versions of it if one want the binary cache to be available both on usual and GCE installations.